Android can revoke file permissions at inconvenient times

If you're unlucky Android will revoke the read permission from the file you're trying to send before/during an active file transfer.

```
E/FileTransferManager: Setting 1234ABCD 0 to status RESUME
E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-2
    Process: ltd.evilcorp.atox.debug, PID: 5509
    java.lang.SecurityException: Permission Denial: opening provider com.android.providers.media.MediaDocumentsProvider from ProcessRecord{9189b12 5509:ltd.evilcorp.atox.debug/u0a153} (pid=5509, uid=10153) requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2373)
        at android.os.Parcel.createException(Parcel.java:2357)
        at android.os.Parcel.readException(Parcel.java:2340)
        at android.os.Parcel.readException(Parcel.java:2282)
        at android.app.IActivityManager$Stub$Proxy.getContentProvider(IActivityManager.java:5702)
        at java.lang.reflect.Method.invoke(Native Method)
        at leakcanary.ServiceWatcher$install$4$2.invoke(ServiceWatcher.kt:85)
        at java.lang.reflect.Proxy.invoke(Proxy.java:1006)
        at $Proxy2.getContentProvider(Unknown Source)
        at android.app.ActivityThread.acquireProvider(ActivityThread.java:6813)
        at android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider(ContextImpl.java:2935)
        at android.content.ContentResolver.acquireUnstableProvider(ContentResolver.java:2481)
        at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1967)
        at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:1798)
        at android.content.ContentResolver.openInputStream(ContentResolver.java:1475)
        at ltd.evilcorp.domain.feature.FileTransferManager.sendChunk(FileTransferManager.kt:240)
        at ltd.evilcorp.atox.tox.EventListenerCallbacks$setUp$$inlined$with$lambda$11.invoke(EventListenerCallbacks.kt:154)
        at ltd.evilcorp.atox.tox.EventListenerCallbacks$setUp$$inlined$with$lambda$11.invoke(EventListenerCallbacks.kt:50)
        at ltd.evilcorp.domain.tox.ToxEventListener.fileChunkRequest(ToxEventListener.kt:119)
        at ltd.evilcorp.domain.tox.ToxEventListener.fileChunkRequest(ToxEventListener.kt:33)
        at im.tox.tox4j.impl.jni.ToxCoreEventDispatch$$anonfun$dispatchFileChunkRequest$1.apply(ToxCoreEventDispatch.scala:162)
        at im.tox.tox4j.impl.jni.ToxCoreEventDispatch$$anonfun$dispatchFileChunkRequest$1.apply(ToxCoreEventDispatch.scala:155)
        at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
        at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
        at scala.collection.Iterator$class.foreach(Iterator.scala:891)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
        at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
        at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
        at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
        at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
        at im.tox.tox4j.impl.jni.ToxCoreEventDispatch$.dispatchFileChunkRequest(ToxCoreEventDispatch.scala:155)
        at im.tox.tox4j.impl.jni.ToxCoreEventDispatch$.dispatchEvents(ToxCoreEventDispatch.scala:223)
        at im.tox.tox4j.impl.jni.ToxCoreEventDispatch$.dispatch(ToxCoreEventDispatch.scala:240)
        at im.tox.tox4j.impl.jni.ToxCoreImpl.iterate(ToxCoreImpl.scala:155)
        at ltd.evilcorp.domain.tox.ToxWrapper.iterate(ToxWrapper.kt:50)
        at ltd.evilcorp.domain.tox.Tox$start$3$1.invokeSuspend(Tox.kt:79)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.am.ActivityManagerService.getContentProviderImpl(ActivityManagerService.java:7155)
        at com.android.server.am.ActivityManagerService.getContentProvider(ActivityManagerService.java:7594)
        at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2381)
        at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:2883)
        at android.os.Binder.execTransactInternal(Binder.java:1154)
```

--- a/atox/src/main/kotlin/ui/chat/ChatFragment.kt
+++ b/atox/src/main/kotlin/ui/chat/ChatFragment.kt
@@ -4,6 +4,7 @@ import android.app.AlertDialog
 import android.content.ActivityNotFoundException
 import android.content.ClipData
 import android.content.ClipboardManager
+import android.content.Context
 import android.content.Intent
 import android.net.Uri
 import android.os.Bundle
@@ -50,6 +51,13 @@ import ltd.evilcorp.domain.tox.PublicKey
 const val CONTACT_PUBLIC_KEY = "publicKey"
 private const val MAX_CONFIRM_DELETE_STRING_LENGTH = 20
 
+class OpenMultiplePersistableDocuments : ActivityResultContracts.OpenMultipleDocuments() {
+    override fun createIntent(context: Context, input: Array<out String>): Intent {
+        return super.createIntent(context, input)
+            .addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION)
+    }
+}
+
 class ChatFragment : BaseFragment<FragmentChatBinding>(FragmentChatBinding::inflate) {
     private val viewModel: ChatViewModel by viewModels { vmFactory }
 
@@ -64,9 +72,10 @@ class ChatFragment : BaseFragment<FragmentChatBinding>(FragmentChatBinding::infl
     }
 
     private val attachFilesLauncher =
-        registerForActivityResult(ActivityResultContracts.OpenMultipleDocuments()) { files ->
+        registerForActivityResult(OpenMultiplePersistableDocuments()) { files ->
             viewModel.setActiveChat(PublicKey(contactPubKey))
             for (file in files) {
+                activity?.contentResolver?.takePersistableUriPermission(file, Intent.FLAG_GRANT_READ_URI_PERMISSION)
                 viewModel.createFt(file)
             }
         }

--- a/domain/src/main/kotlin/feature/FileTransferManager.kt
+++ b/domain/src/main/kotlin/feature/FileTransferManager.kt
@@ -2,6 +2,7 @@ package ltd.evilcorp.domain.feature
 
 import android.content.ContentResolver
 import android.content.Context
+import android.content.Intent
 import android.net.Uri
 import android.provider.OpenableColumns
 import android.util.Log
@@ -49,6 +50,10 @@ class FileTransferManager @Inject constructor(
     init {
         File(context.filesDir, "ft").mkdir()
         File(context.filesDir, "avatar").mkdir()
+        resolver.persistedUriPermissions.forEach {
+            Log.w(TAG, "Clearing leftover permission for ${it.uri}")
+            resolver.releasePersistableUriPermission(it.uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
+        }
     }
 
     fun reset() {
@@ -64,6 +69,10 @@ class FileTransferManager @Inject constructor(
             setProgress(ft, FtRejected)
             fileTransfers.remove(ft)
             File(ft.destination).delete()
+            if (ft.outgoing) {
+                val uri = Uri.parse(ft.destination)
+                resolver.releasePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
+            }
         }
     }
 
@@ -140,7 +149,11 @@ class FileTransferManager @Inject constructor(
         fileTransfers.remove(ft)
         setProgress(ft, FtRejected)
         tox.stopFileTransfer(PublicKey(ft.publicKey), ft.fileNumber)
-        File(Uri.parse(ft.destination).path!!).delete()
+        val uri = Uri.parse(ft.destination)
+        File(uri.path!!).delete()
+        if (ft.outgoing) {
+            resolver.releasePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
+        }
     }
 
     private fun setDestination(ft: FileTransfer, destination: Uri) {
@@ -225,13 +238,14 @@ class FileTransferManager @Inject constructor(
             return
         }
 
+        val src = Uri.parse(ft.destination)
         if (length == 0) {
             Log.i(TAG, "Finished outgoing ft ${pk.take(8)} $fileNo ${ft.isComplete()}")
             fileTransfers.remove(ft)
+            resolver.releasePersistableUriPermission(src, Intent.FLAG_GRANT_READ_URI_PERMISSION)
             return
         }
 
-        val src = Uri.parse(ft.destination)
         val bytes = resolver.openInputStream(src)?.use {
             it.skip(pos)
             val bytes = ByteArray(length)

--- a/domain/src/main/kotlin/feature/FileTransferManager.kt
+++ b/domain/src/main/kotlin/feature/FileTransferManager.kt
@@ -68,10 +68,11 @@ class FileTransferManager @Inject constructor(
         fileTransfers.filter { it.publicKey == pk }.forEach { ft ->
             setProgress(ft, FtRejected)
             fileTransfers.remove(ft)
-            File(ft.destination).delete()
             if (ft.outgoing) {
                 val uri = Uri.parse(ft.destination)
                 resolver.releasePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
+            } else {
+                File(ft.destination).delete()
             }
         }
     }
@@ -150,9 +151,10 @@ class FileTransferManager @Inject constructor(
         setProgress(ft, FtRejected)
         tox.stopFileTransfer(PublicKey(ft.publicKey), ft.fileNumber)
         val uri = Uri.parse(ft.destination)
-        File(uri.path!!).delete()
         if (ft.outgoing) {
             resolver.releasePersistableUriPermission(uri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
+        } else {
+            File(uri.path!!).delete()
         }
     }
 

--- a/domain/src/main/kotlin/feature/FileTransferManager.kt
+++ b/domain/src/main/kotlin/feature/FileTransferManager.kt
@@ -248,14 +248,18 @@ class FileTransferManager @Inject constructor(
             return
         }
 
-        val bytes = resolver.openInputStream(src)?.use {
-            it.skip(pos)
-            val bytes = ByteArray(length)
-            it.read(bytes, 0, length)
-            bytes
-        } ?: return
-        tox.sendFileChunk(PublicKey(pk), fileNo, pos, bytes)
-        setProgress(ft, ft.progress + length)
+        try {
+            val bytes = resolver.openInputStream(src)?.use {
+                it.skip(pos)
+                val bytes = ByteArray(length)
+                it.read(bytes, 0, length)
+                bytes
+            } ?: return
+            tox.sendFileChunk(PublicKey(pk), fileNo, pos, bytes)
+            setProgress(ft, ft.progress + length)
+        } catch (e: SecurityException) {
+            Log.e(TAG, "sendChunk: $e")
+        }
     }
 
     fun setStatus(pk: String, fileNo: Int, fileStatus: ToxFileControl) {

